### PR TITLE
Add parallel only checks

### PR DESCRIPTION
**Pull request by lindsayad**
Defensive programming to help protect MOOSE

Refs hangs associated with ncrc/griffin#1481

### CODE REVIEW

File: framework/include/problems/FEProblemBase.h [hunk 1]
```diff
@@ -2590,6 +2590,8 @@ FEProblemBase::addObject(const std::string & type,
                          InputParameters & parameters,
                          const bool threaded)
 {
+  libmesh_parallel_only(this->comm());
```
Comment (roystgnr) [inline framework/include/problems/FEProblemBase.h:2593]: Would parallel_object_only() work here?

**Review by loganharbour — APPROVED** [review]

**Review by roystgnr — APPROVED** [review]

File: framework/include/problems/FEProblemBase.h [hunk 2]
```diff
@@ -2590,6 +2590,8 @@ FEProblemBase::addObject(const std::string & type,
                          InputParameters & parameters,
                          const bool threaded)
 {
+  parallel_object_only();
```
Comment (GiudGiud) [inline framework/include/problems/FEProblemBase.h:2593]: what does this outut

**Comment by roystgnr** [comment]
The current message is `Different ranks are at different libmesh_parallel_only points`.  I guess we could change that to `Entering a parallel section, but not in parallel`?

> One of your processes has crashed/frzoen before this point

This is somewhat misleading, though.  When two processes take two different branches by mistake, which of them is the "frozen" one?  Only the combination could be called "frozen", with each expecting the other to be elsewhere.

**Comment by loganharbour** [comment]
This could be made better in a few ways, that's true.

However, it's an assertion check and would have saved @lindsayad more than a day of debugging. This makes it a clear win, with room for improvement. IMO the sooner this gets in, the better.

**Comment by roystgnr** [comment]
> This is a developer check, not a user check.

I'm afraid it ought to stay that way, too, despite how confusing it can be for a run to go out of sync without these checks.  Parallel communication is too expensive to want to do blocking reductions any more often than we have to - especially for cases like some of the calls in this PR where we're not layering the check on top of an existing communication, we're just making sure that some data stays in sync lest we get bitten by a discrepancy much later.

**Comment by GiudGiud** [comment]
> This is somewhat misleading, though. When two processes take two different branches by mistake, which of them is the "frozen" one?

This is true. I guess the message would have to say:

- You are doing this particular item for which we have an assert wrong
OR
- Rank xxxx froze before this for some reason (can we send the backtrace for the divergent rank? Or too tough from C++?)
OR
- Rank xxxx is on a different code path, waiting at this "other" parallel section (likely another parallel_only) (can we detect that?)

**Comment by loganharbour** [comment]
> I honestly don't recommend anyone build complex simulations in opt mode

This is a moot point for many use cases. Anyone that has application binary access does not have access to a binary with symbols or assertions. 

My general philosophy is very much put developer errors (which this one is) as assertions, user errors as standard errors. This exposes a bigger problem - better user diagnostics. But I don't think that problem is related to this one.

**Comment by loganharbour** [comment]
> Why would adding rather unclear asserts not be related to providing users with more insight about what's going on when the code crashes?

Because our time can be spent more efficiently on more common issues and diagnostics

**Comment by roystgnr** [comment]
I've been trying hard to hold my tongue, not beat a dead horse (in an inappropriate thread! what is the best thread to debate development standards?), and not cause a flame war inadvertently ... but I lack self-control, so here we are.  Maybe I can still pull off the last out of the three?  Please interpret my following comments with more friendliness than I know how to convey over ASCII...

> I honestly don't recommend anyone build complex simulations in opt mode.

Agreed (while developing).  I never understood why the default MOOSE build `$METHODS` used to *omit* `devel`, which IMHO ought to be the default way to develop new code.  `opt` isn't safe enough until you've tested something on every feature you intend to use in practice (and ideally every combination of features; damn curse of dimensionality)  `dbg` isn't nearly fast enough; `devel` is slower than I'd like, but `dbg` is not only much slower than that, it's *asymptotically* slower even.  "Hammer all your feature combinations in `devel`, if anything screams switch to `dbg` to figure it out, then when nothing screams switch to `opt` for big problems" just seems like the only development method that makes sense.

> You're missing too much checking.

Specifically, you're missing 900 `mooseAssert` checks out of 142K lines of framework code, and 5600 `libmesh_assert` checks out of 95K lines of libMesh code.

I don't think it's a coincidence that "the standard of MOOSE developer checks" led to an *order of magnitude* lower density of checks.  I think it's inevitable that, when you tell people "every time you do X you need to do Y", and typically Y is *impossible*, people will do less X.

Why do I say adding a user-friendly message to an assertion is typically impossible?

1. Theoretically: an assertion is something we add for a precondition or postcondition that we expect should *never* be violated in merged code.  It's the equivalent of the `_GLIBCXX_DEBUG` thing that can force every single `vector::operator[]` call to do range-checking.  Arguably the only true solution (and definitely the only performant solution) is "never access a vector outside its range", just like it's "never call a MOOSE or libMesh function with invalid arguments or invalid global state".  It's certainly possible for any code that's not certain it's doing the right thing to make the same tests itself in those cases.  But sometimes someone screws up.  The trouble is that, by definition it's a developer or developers have screwed up here: if a user mistake can lead to a condition being unmet, that should be a `libmesh_error_msg` or `mooseError`, enabled in opt mode with a user-friendly message; assertions are for the cases where that *isn't* true.  So in theory failed assertions are things that a user didn't cause and can't fix.  The most important thing is just to give them *some* message, with file and line numbers, to pass along to a developer to fix.

2. Empirically: just `git grep -A2 mooseAssert` and see.  Some of these should be redundant with `mooseError` calls elsewhere, and I do strongly approve of the redundancy, and using a redundant error message here is a good idea too ... but look at the few places where someone dared to add pure internal consistency checks.  What is a user supposed to do after being told "qrule"/"elem"/etc "must be non-null", or "This has to match something", or "Wrong size for dof_indices"?  

To use the current example: "we tried to add an object on all ranks, but it failed. One or more of your ranks is not reaching this" is a really good error message ... but it still debateably *isn't* a better error message than even "assertion failed at FEProblemBase.h line 2593" would have been.  From the user side, being given some friendly-*sounding* message about what's going wrong [falsely suggests](https://en.wikipedia.org/wiki/Cooperative_principle) that the user ought to be paying attention to the message and trying to figure out how their input file caused it, rather than just passing along a file name and line number to developers ASAP.  From the app developer side, figuring out what's going on from "Different ranks are at different libmesh_parallel_only points" plus file names plus line numbers is the *easy* part of the problem; if that's still not easy enough then "we'll add more details so you can start parallel debugging without getting help" would not be a kindness.  Finally, from the framework developer side, anyone who wants to come up with an actually-user-friendly way to explain something like this (or, say, "Wrong size for dof_indices") is going to come up short, and so has to choose between either adding a user-baffling message (and so still violating the MOOSE standard in spirit, if not in letter) or adding no assertion at all.  I think I've taken both options, myself.  My guess is that there have been approximately `5600/95e3*142e3 - 900` cases, several thousand, where people consciously or unconsciously made the latter choice.

But then, if we're not going to be adding 90% of the devel+dbg-mode assertions we should be adding to begin with, at some point we might as well just tell everybody to build in opt mode.  If we're lucky they'll hit a non-assertion error or an MPI hang or [nasal demons](http://www.catb.org/jargon/html/N/nasal-demons.html), and in the worst case at least they'll get to the wrong answer faster.

**Comment by lindsayad** [comment]
> Why would adding rather unclear asserts not be related to providing users with more insight about what's going on when the code crashes?

If I were to add an error message to this, it would be something like "boundary condition 'type/name' not being added on all processes". This represents an improvement over the current message if it's more helpful to the developer that the user reports the problem to. The user will have no immediate recourse for fixing the problem without developer intervention. And for myself, an additional message would be of no more use than the current file and line number information that is already output. That was all the help I needed to identify the problem.

> I honestly don't recommend anyone build complex simulations in opt mode

I think that's fair. But I think the solution for this is to do more testing of our applications in `devel/dbg` modes. The `libmesh_parallel_only` check is a global communication that if you have a lot of objects can really slow down simulation setup. I experienced this with @YaqiWang's big problem. Adding the `devel` checks really slowed things down and I wasn't able to make meaningful debugging progress until he was able to reproduce with a smaller problem.

**Comment by GiudGiud** [comment]
> Agreed (while developing). I never understood  ... etc

Agree though I never think these things as thoroughly as you do.

> an order of magnitude lower density of checks.

I don't think (helpful messages being impossible) that's the reason for that. Being more on the user facing side in MOOSE means we tend to think about user errors more than development errors. In libmesh the user is very likely to be a developer. We're incentivized to protect users more than we are trying to protect devs, so when finalizing our code we make sure to add the mooseError we should have, and don't do a second pass at adding mooseAsserts

in fact there's 1800 mooseErrrors, twice the asserts. And there probably should be more still

I would also not say that we have been imposing helpful assert messages, hence why I 'm asking for it. So not having imposed that standard, we can't blame it for the thousands of missing asserts when comparing MOOSE to libmesh. 

>  "Wrong size for dof_indices"

On paper here it's not helpful. in practice:
- if the dev already knows why this could be, then adding that information to the assert would help the other dev that has to fix it. If this boils down to something like the order of execution of auxkernels and UOs for example (UOs initializing something) then a user could self-help on a message like this. Does it ought to be an error instead? Yes probably. 

- it looks better to have a reason for the crash rather than no information, just a line number. "it crashed at line 2353 of this file" looks like broken code. I m even going to need your commit number to debug it, as this line could have moved. "it crashed because the conditions for this kernel were not met as evidenced by the error message referring to this data structure that is not the right size", that looks like we did defense in depth and saved the user from executing in bad conditions. Even though it s nearly the same code!

>  if that's still not easy enough then "we'll add more details so you can start parallel debugging without getting help

We want them to try to debug what s going on. We dont have the resources to debug everything for everyone, even though that's what we do and at the expense of our own research. 

>  so has to choose between either adding a user-baffling message (and so still violating the MOOSE standard in spirit, if not in letter) or adding no assertion at all.

in this theoretical world where we dont accept PRs that add asserts that are not giving much context about what's going on. We dont do that. I just want context when we can. After all, we are crashing the simulation, there is nothing after that point. The user is stuck, they need to be unstuck. So every lead we think we have on why this could be can help them keep moving. 

> But then, if we're not going to be adding 90% of the devel+dbg-mode assertions we should be adding to begin with, at some point we might as well just tell everybody to build in opt mode. If we're lucky they'll hit a non-assertion error or an MPI hang or [nasal demons](http://www.catb.org/jargon/html/N/nasal-demons.html), and in the worst case at least they'll get to the wrong answer faster.

We will be adding these asserts. As this PR does. See above, I disagree adding the most context and the most suggestions we can get is preventing that.

**Comment by roystgnr** [comment]
> in fact there's 1800 mooseErrrors, twice the asserts.

Well, this is a good point.  libMesh has more, but properly counting both cases gets the ratio from nearly 10x down to like 4x.  And typical libMesh error messages usually aren't as good as MOOSE ones.  The evil twin of the libMesh "assert every precondition and postcondition" philosophy is "libMesh is user friendly, it's just picky about who its friends are".

> Honestly, even telling the user "run in serial to avoid this!" is an improvement over what we have now

It was only [early this year](https://github.com/libMesh/libmesh/pull/3476) that we got a decent message for these macros at all; I'm open to rewording it.  None of my complaints about the hurdle of writing a friendly invocation-specific error message apply to improving the general reusable messages that we output.

Actually, speaking of the output ...

> One or more (and maybe we should output which ones) of your ranks is not reaching this

What's a default MOOSE build doing when a `libmesh_terminate_handler()` makes it to `write_traceout()`?  I see we're configuring `--without-gdb-command` (which is reasonable; gdb seems inordinately slow these days), so we're not going to get fancy gdb-parsed backtraces, and I don't think we have any other way of getting stack traces on OSX, but are we at least getting `traceout_` (MPI rank here) `_` (Unix pid here) `.txt` files with glibc-parsed backtraces on Linux?  I see them in my own MOOSE test directories but that may just be because of runs built against my own libMesh configurations.

**Comment by lindsayad** [comment]
MOOSE does not configure with `--enable-tracefiles`. However, even without them libMesh does write the process in `here`:
```c++
void here(const char * file, int line, const char * date, const char * time, std::ostream & os)
{
  os << "[" << static_cast<std::size_t>(libMesh::global_processor_id()) << "] "
     << file
     << ", line " << line
     << ", compiled " << date
     << " at " << time
     << std::endl;
}
```
so the result is that `parallel_object_only()` does indicate exactly (in the console output) how many processes and which of them end up in one location vs. another.

**Comment by roystgnr** [comment]
That's better than nothing, I guess.  I do generally find having at least one extra level of a stack trace is often a lot more helpful than only having the specific spot where the discrepancy was detected, though.

I could write much more clever diagnostics that would work in the "everybody hits a `parallel_object_only()` type macro from different places" case ... but we'd really need to get more zealous about adding those macros before that would be useful.  If 255 ranks hit the allreduce calls from a `parallel_object_only()` while 1 rank hits a non-allreduce MPI call, IIRC the MPI standard doesn't actually say what should happen, and in practice the answer is "nothing good".  Experimenting with MPICH, if I have one rank hit a sum-of-maps (just the first thing in a long unit test list) at the same time a second hits a parallel_only assertion, they both hang in a busy-loop, the former in an allgather alongside the latter in the allreduce.  I don't see how to do anything robustly clever here at lower levels without being so intrusive that it would cut our `devel` mode communication speed in half.

**Comment by lindsayad** [comment]
We could consider adding the `--enable-tracefiles` option to libMesh configure (I myself always configure with that option). I believe @friedmud and @permcody told me that we did at one point but that users got tired of hundreds of `.txt` files showing up in their HPC directories. But I could be misremembering

**Comment by roystgnr** [comment]
Well, that is something that it's easy to get tired of.  I wonder if there's anything simple to be done (put all trace files from a particular crash into one subdirectory, so you get `N_crashes` entries at the top level instead of `N_crashes*N_ranks`?) that might make it less annoying.

**Comment by roystgnr** [comment]
Huh.  Yeah.  It's silly to make that a compile-time option only.

**Pull request merged**